test/fuzz/umfFuzz.cpp

```diff
@@ -0,0 +1,191 @@
+    }
+    test_state.allocated_memory.push_back(std::make_pair(ptr, alloc_size));
+    std::cout << "Allocated memory at " << ptr << " with alloc_size "
+              << alloc_size << std::endl;
+    std::cout << "Vector size is: " << test_state.allocated_memory.size()
+              << std::endl;
+    return 0;
+}
+
+int umf_memory_provider_free(TestState &test_state) {
+    printf("begin umf_memory_provider_free\n");
+    if (test_state.allocated_memory.empty()) {
+        return -1;
+    }
+
+    std::pair<void *, size_t> alloc = test_state.allocated_memory.back();
+    umf_result_t res =
+        umfMemoryProviderFree(test_state.provider, alloc.first, alloc.second);
+
+    if (res != UMF_RESULT_SUCCESS) {
+        std::cout << "Failed to free memory to the provider! " << res
+                  << std::endl;
+        ;
+        return -1;
+    }
+
+    std::cout << "Freed memory at " << alloc.first << " with alloc_size "
+              << alloc.second << std::endl;
+    test_state.allocated_memory.pop_back();
+    return 0;
+}
+
+int umf_pool_create(TestState &test_state) {
+    if (test_state.pools.size() > MAX_POOLS_VECTOR_SIZE) {
+        return -1;
+    }
+
+    umf_memory_pool_ops_t *pool_ops = umfScalablePoolOps();
+    void *pool_params = NULL;
+    umf_pool_create_flags_t flags = 0;
+    umf_memory_pool_handle_t pool;
+    umf_result_t res =
+        umfPoolCreate(pool_ops, test_state.provider, pool_params, flags, &pool);
+
+    if (res != UMF_RESULT_SUCCESS) {
+        std::cout << "Failed to create a pool! " << res << std::endl;
+        return -1;
+    }
+
+    test_state.pools.push_back(pool);
+    std::cout << "Scalable memory pool created at " << pool
+              << "and Pools' vector size is: " << test_state.pools.size()
+              << std::endl;
+    return 0;
+}
+
+int umf_pool_destroy(TestState &test_state) {
+    if (test_state.pools.empty()) {
+        return -1;
+    }
+    umfPoolDestroy(test_state.pools.back());
+    test_state.pools.pop_back();
+    return 0;
+}
+
+int umf_pool_malloc(TestState &test_state) {
+    if (test_state.pools.empty()) {
+        return -1;
+    }
+    size_t alloc_size;
+    int ret = get_alloc_size(test_state, alloc_size, MAX_POOLS_ALLOC_SIZE);
+    if (ret != 0) {
+        return -1;
+    }
+
+    void *ptr = umfPoolMalloc(test_state.pools.back(), alloc_size);
+    if (!ptr) {
+        std::cout << "Failed to allocate memory in the pool!" << std::endl;
+        return -1;
+    }
+    test_state.pools_alloc.push_back(ptr);
+    return 0;
+}
+
+int umf_free(TestState &test_state) {
+    if (test_state.pools_alloc.empty()) {
+        return -1;
+    }
+    umfFree(test_state.pools_alloc.back());
+    test_state.pools_alloc.pop_back();
+    return 0;
+}
+
+void cleanup(TestState &test_state) {
+    for (auto &alloc : test_state.allocated_memory) {
+        umfMemoryProviderFree(test_state.provider, alloc.first, alloc.second);
+    }
+    std::cout << "Freed all allocated memory\n";
```

Comment:
Memory from the provider or from pools? Be specific and add more messages.

Reply:
done